### ISSUE TITLE
[TritonIntelGPU] Enable `CanonicalizePointers` pass by default back

### ISSUE THREAD

**Issue by anmyachev** [intel/intel-xpu-backend-for-triton#7540]
Was disabled by default in https://github.com/intel/intel-xpu-backend-for-triton/pull/7532 to fix wide performance regressions across E2E models in https://github.com/intel/intel-xpu-backend-for-triton/issues/7511

**Comment by whitneywhtsang** [comment]
## gemm-grouped-batched regression confirmed caused by disabling CanonicalizePointers

The `gemm-grouped-batched` benchmark regression (12.2 → 10.7 TFlops GeoMean, ~12.5% drop) between the July 18 and July 22 CI runs on PVC is caused by commit 60fe135f1 (disable `CanonicalizePointers` by default).

### Local reproduction on Intel Max 1100

| Shape (G-M-N-K) | CP Disabled (default) | CP Enabled | Change |
|---|---|---|---|
| 8-512-4096-4096 | 8.46 TFlops | 10.13 TFlops | +20% |
| 8-1024-4096-4096 | 8.34 | 9.81 | +18% |
| 16-512-2048-2048 | 8.38 | 10.25 | +22% |
| 16-1024-4096-4096 | 8.53 | 10.03 | +18% |
| 32-256-4096-4096 | 8.53 | 10.14 | +19% |
| 64-128-2048-2048 | **9.78** | 9.15 | **-7%** |
| 128-64-1024-1024 | 9.26 | 10.06 | +9% |

Re-enabling via `TRITON_INTEL_DISABLE_CANONICALIZE_POINTERS=0` recovers 18-22% on the larger shapes.

The exception is `[64-128-2048-2048]` which regresses 7% with CP enabled — this is likely the class of regression that motivated disabling CP in #7515.

### Context

This benchmark uses PyTorch Inductor's `grouped_mm` Triton template with pointer-based loads (`USE_TMA_LOAD=False`). CanonicalizePointers helps the Coalesce and MaterializeBlockPointer passes analyze pointer arithmetic patterns generated by Inductor, enabling more efficient memory access layouts.

<img width="613" height="379" alt="Image" src="https://github.com/user-attachments/assets/15324da4-7ee8-4641-a1f0-fd3cd988dbfb" />

**Comment by etiotto** [comment]
## Root cause of the CP-ON regression

The blocker is a **cost-model mispricing in `RemoveLayoutConversions`**, not a defect in `CanonicalizePointers`. CP only exposes it.

`RemoveLayoutConversions` decides whether to delete a layout conversion by rematerializing the value in the target layout instead. It is a single judgment: *is the conversion I'd delete more expensive than the work I'd have to duplicate?* The two sides of that comparison were priced on different terms.

**The conversion side was priced as if every convert goes through shared local memory.** SLM has 32 banks and moves 32-bit units, so `getConvertCost` rounds every tensor up to at least 32 elements of at least 32 bits before costing it (`getByteCount(src, /*minElementCount=*/32, /*minBitWidth=*/32)`). In the LayoutLM attention softmax the convert is on a **single f32** — the result of a row-wise `max` — and it never touches SLM at all; it lowers to a few register moves. The `-intel-allocate-shared-memory` pass assigns it **0 bytes**, and lowering it in isolation yields 9 LLVM instructions with 0 shuffles and 0 `addrspace(3)`. The rounding priced that nearly-free convert **32x too high**: 12288 instead of 384.

**The rematerialization side is priced honestly**, from real tensor sizes with no clamp. What it duplicates is a 1x512 global load plus a full reduction: 10304.

Inflated number vs. honest number, so `convertLayoutCost >= rematerialisationCost` concluded the convert was the expensive one, and the pass duplicated a global load and a reduction in order to remove some register moves.

### Why it only surfaces with CP enabled

The margin was 64 bytes wide. Verified by replaying the exact RLC#2 input module (pass 30, after `CanonicalizePointers`@25 and `MaterializeBlockPointer`@29) through `triton-opt` for both configurations:

| | convert cost | remat cost | outcome |
|---|---|---|---|
| CP off | 12288 | 12352 | barely loses -> correctly skipped |
| CP on | 12288 | 10304 | wins -> remat fires |

CP splits fat pointers into a scalar base plus a tensor offset, which puts one extra producer into the backward slice being duplicated and shifts its cost down by one 1x512xf32 (2048 bytes) — enough to flip the decision. The mispricing was already there, sitting just under the threshold.

### The damage

Once that remat fires, the softmax kernel is left in **two co-resident blocked layouts**: the input row is loaded twice, the whole numerator chain (`mulf`/`absf`/`cmpf`/`tt.reduce`) is duplicated (`tt.reduce` 8 -> 14), RLC folds only 3 of 4 converts instead of collapsing to one layout, and the duplicated load gets relabeled `ttig.block_io = "column_major"` — a strided copy of a row-major read.

- LLIR on each of the two hot per-layer softmax kernels: **1064 -> 8354 lines (7.9x)**
- LayoutLMForMaskedLM bf16 training on PVC 1100 (Max 1100), dynamo harness: **99.8 ms -> 172.7 ms (1.73x slower)**

The regression is localized to exactly these 2 kernels. All 34 other differing kernels are flat (<=1.04x), and the pointwise transpose kernels actually **improve** — CP correctly hoists the uniform offset to a scalar `addptr`. The `ttig.block_io` fast-path count is essentially unchanged (195 -> 197), so CP does not break 2D block I/O; the damage is entirely the RLC layout fixpoint.

Note that the existing guard intended to catch this class of damage — `rematDegradesLoad`, the 2D-block-I/O veto from #7080 — does not fire here, because it applies a strict `>` on `estimateLoadHWCost` and that rates row-major and column-major block loads as equal.

This is the same failure class as the RLC remat/layout-fixpoint coupling seen previously: convert *count* is not a performance proxy, and a small change in one remat decision can flip the global layout fixpoint for a whole kernel.

**Comment by etiotto** [comment]
## The fix

The convert cost estimator rounds every tensor up to at least **32 elements** and at least **32 bits** before pricing it (`getByteCount(src, /*minElementCount=*/32, /*minBitWidth=*/32)`). Those two roundings were bundled together, but they exist for different reasons:

- **32 elements** exists only because SLM has 32 banks and stores one element per bank. It is an artifact of going through shared memory.
- **32 bits** exists because a register is 32 bits wide — an `i1` still takes a whole one.

A register-only convert never goes through shared memory, so the first rounding describes something that is not happening. That single rounding is what turned a 1-element convert into a 32-element one and made it look 32x more expensive than it is.

So: apply the element rounding only when the convert actually uses shared memory. Keep the bit rounding always.

```cpp
int64_t minElementCount = cvtReordersRegisters(srcType, resultType) ? 0 : 32;
int64_t convertLayoutBytes = getByteCount(convertSrc, minElementCount, 32);
```

Nothing else changes — the floor still applies to real SLM converts, and the Intel `*3` SLM-sync factor (#5476) is untouched. The signature becomes `getConvertCost(Value convertSrc, Attribute resultEncoding)`, matching upstream, and the two call sites are updated. 23 lines.

## Why not just price free converts at zero

That is what upstream does (`if (cvtReordersRegisters(srcTy, resultType)) return 0;` in `lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp`), and it is wrong here.

The decision is `convertCost >= rematCost`. Setting `convertCost` to 0 makes it `0 >= rematCost`, which is true **only when `rematCost` is 0 too**. So a free convert can no longer justify *any* remat that costs something, however cheap. `combine.mlir @reduce_cvt2` has a register-only convert whose remat genuinely is profitable (`rematCost` 1088), and zeroing the cost suppresses it. Upstream does not hit this with its own same-named test because its layouts there make `cvtReordersRegisters` false, so the early-out never applies.

Keeping the cost proportional to real tensor size means **size** does the discriminating, which is what you actually want:

| Convert | Cost | Remat cost | Decision |
|---|---|---|---|
| 1 element f32 (the bug) | 384 | 10304 | don't remat :white_check_mark: |
| 256 element i1 (`@reduce_cvt2`) | 98304 | 1024 | remat :white_check_mark: |

One rule, correct on both.

## Effect

Replaying the RLC#2 input module for the LayoutLM softmax now gives `loads=2 reduce=4 cvt=4` with CP enabled — byte-identical to the CP-disabled reference. End to end on PVC 1100 (Max 1100), LayoutLMForMaskedLM bf16 training:

| Config | `abs_latency` | speedup vs eager |
|---|---|---|
| CP off (current default) | 99.77 ms | 1.615x |
| CP on, before fix | 172.71 ms | 0.926x |
| **CP on, after fix** | **98.77 ms** | **1.618x** |

Both hot softmax kernels are back to 1065/1066 lines of LLIR (from 8354), and no kernel in the run exceeds 3000 lines. Lit is green: 76/76 `test/TritonIntelGPU/` and 130/130 upstream `test/TritonGPU/`, plus a new regression test `test/TritonIntelGPU/RemoveLayoutConversions/remat_free_convert_cost.mlir` that fails without the fix.

Note on scope: this validates the fix on the one model I reproduced. Flipping the `knobs.py` CP default still requires measuring the ~23 models originally regressed in #7511, so I intend to keep that as a separate change.

**Comment by etiotto** [comment]
PR #7685 posts the fix described above (RLC convert-cost mispricing only; the CP default flip stays with this issue).

Scope caveat for the record: the 8 rows that recover on PVC are all LayoutLM. The Bert / Roberta / Electra / Megatron rows from #7511 (1.43-1.68x on BMG) are **flat on PVC before the fix** (0.98-1.05x), so they have a separate cause that this change neither helps nor hurts. `mobilevit_s` (3 rows) and `hf_T5_large` (1 row) are unmeasured here — no `timm` or torchbench in my environment. Flipping the CP default therefore still needs a BMG run of that cluster.
